### PR TITLE
Core elements of building and running Docker image

### PR DESCRIPTION
**Pull request by ijdickinson**
This PR proposes some of the key elements for gettig the UKHPI
application to run in a Docker container, including:

- a Makefile to build images and run images in a local container
- a Dockerfile that specifies the image
- an `entrypoint.sh` that performs some checks before starting Rails
- migration of some key parameters, such as the URL of the API service,
  to environment variables
- logging to stdout if a particular environment variable is set

Given the discussion on [another Github
issue](https://github.com/epimorphics/hmlr-infrastructure-migration/issues/1),
we assume that the DSAPI is available to the Docker image via a mapped
port. However, I have been unable to get this to work, except by using
`host` networking. That's not ideal, and should be revisited by someone
who is better versed in Docker networking than I am.

There will be more to sort out regarding publishing images to Docker
repositories and production environments, so the only `Makefile` targets
I have tested at this point are `image` and `prod`. The other targets
were copied from another `Makefile`, and doubtless will need more work.

GitHub issue #239

### CODE REVIEW

File: config/environments/production.rb [hunk 1]
```diff
@@ -80,5 +80,15 @@
   # Disable logging of the rendering of partials
   config.action_view.logger = nil
 
-  config.relative_url_root = '/app/ukhpi'
+  if ENV['RAILS_LOG_TO_STDOUT'].present?
+    logger           = ActiveSupport::Logger.new(STDOUT)
+    logger.formatter = config.log_formatter
```
Comment (der) [inline config/environments/production.rb:85]: Not sure what this formatter does but it would nice to support single line and structured (json) logs in production for ease of monitoring. Looking around then `lograge` seems to the the most frequently mentioned gem which does support compact structure logs and offers a builtin json format option so don't have to do that manually.

**Review by der — COMMENTED** [review]
Looks good.

Apart from the embedded comment on logging my main comment is that this is running the built in rails server raw, so there's no threading or multi-process support. Don't think that would be viable in production and especially not for LR.

Of the rails-compatible app servers we've used passenger up till now. It has been a pain to install and keep up to date but with the docker approach that would get simplified. Looking round at alternatives it seems to be a choice between puma, unicorn and passenger. Unicorn has issues, though the front end nginx could be set up to defray those (e.g. buffering of slow clients). Random googling suggests puma would be a good default choice and, as I understand it, is what heroku have switched to. Maybe should test that.

Whether configuring the app server should be part of this PR or a further one I leave up to you. If you'd like to get his checked and come back to the app server issue that I'd be OK with that. But eventually that'll need to part of this Docker image.

tl;dr needs an app server, now or later

File: config/environments/production.rb (hunk 1 above)
Comment (ijdickinson) [reply config/environments/production.rb:85]: Yes, I was planning to add a JSON log formatter. Added a reminder as  #241

**Comment by ijdickinson** [comment]
@der At the moment, I'm using [puma](https://github.com/puma/puma) with default configuration. There are probably - certainly - options we can tune, but out-of-the box it should support concurrent requests. I think!

**Comment by ijdickinson** [comment]
@der I've added some explicit default Puma configuration, we can fine-tune this over time. Startup now says:

```
docker run --network=host --rm --name ukhpi [...]
=> Booting Puma
=> Rails 6.0.3.2 application starting in production 
=> Run `rails server --help` for more startup options
I, [2020-08-13T17:35:24.755630 #1]  INFO -- : Raven 3.0.0 ready to catch errors
Puma starting in single mode...
* Version 4.3.5 (ruby 2.6.6-p146), codename: Mysterious Traveller
* Min threads: 5, max threads: 5
* Environment: production
* Listening on tcp://0.0.0.0:3000
```

We should look at one of the load-testing tools to see how we might want to tweak the Puma configuration parameters further.

**Comment by der** [comment]
Sorry for not understanding that puma was already in us! Serve me right for reviewing something in a language and framework I know nothing about.

**Review by der — APPROVED** [review]
lgtm

**Review by andrew-pickin-epi — APPROVED** [review]
